### PR TITLE
Fix error unmarshalling for eventstream errors

### PR DESCRIPTION
**Pull request by dagnir**
## Motivation and Context
This fixes an issue in the SDK where we were previously expecting the
`:exception-type` header in the event to be the error code of the
exception that we should unmarshal it to. Instead, we should be mapping
using the member name of the exception within the eventstream structure.

## Testing
 - Added new tests in `codegen` and `protocol-tests`
 - Ran integ tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java [hunk 1]
```diff
@@ -252,6 +269,7 @@ public final SubclassT registerModeledException(ExceptionMetadata errorMetadata)
          * @param exceptionBuilderSupplier Suppplier of the base service exceptions Builder.
          * @return This builder for method chaining.
          */
+        @Deprecated
         public final SubclassT defaultServiceExceptionSupplier(Supplier<SdkPojo> exceptionBuilderSupplier) {
```
Comment (davidh44) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java:273]: can we call out in javadoc its deprecated in favor of new createErrorResponseHandler overload, for future reference

File: test/protocol-tests/pom.xml [hunk 2]
```diff
@@ -255,6 +255,46 @@
...
+            <artifactId>netty-common</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-buffer</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-handler</artifactId>
+        </dependency>
```
Comment (davidh44) [inline test/protocol-tests/pom.xml:281]: should these all be test scope?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java (hunk 1 above)
Comment (dagnir) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java:273]: +1, yes will do!

File: test/protocol-tests/pom.xml (hunk 2 above)
Comment (dagnir) [reply test/protocol-tests/pom.xml]: Yep, done!

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java (hunk 1 above)
Comment (dagnir) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java]: I removed the deprecation for this and deprecated `BaseAseJsonProtocolFactory#createErrorResponseHandler(JsonOperationMetadata)` and `registerModeledException` instead. I'll add a backlog item to remove our usages of these two methods in favor of the new ``BaseAseJsonProtocolFactory#createErrorResponseHandler(JsonOperationMetadata, Function)` introduced here.

Comment (dagnir) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java]: Backlog item created, ref JAVA-7976.

**Review by davidh44 — APPROVED** [review]

**Pull request merged**